### ISSUE TITLE
Tracking: generating completion/manual by running compiled program breaks cross compilation

### ISSUE THREAD

**Issue by Aleksanaa** [NixOS/nixpkgs#308283]
Some programs generates shell completion or manual page with some command line args, e.g. `--completion`, and we usually do this to install completions:

```sh
    installShellCompletion --cmd vrc-get \
      --bash <($out/bin/vrc-get completion bash) \
      --fish <($out/bin/vrc-get completion fish) \
      --zsh <($out/bin/vrc-get completion zsh)
```

However, this process requires executing the resulting binary, which is not possible when running cross compilation.

There should be much more packages out there having this, but they don't do any special handling of cross compilation, so they will very probably fail.

I suggest we add `stdenv.buildPlatform.canExecute stdenv.hostPlatform` toggle, like in other places where we avoid executing in cross compilation, in https://github.com/NixOS/nixpkgs/pull/308055

Another PR https://github.com/NixOS/nixpkgs/pull/307018 built the package targeting `buildPlatform` first, by using `buildPackages.cert-viewer`, which looks good but can cause excessive builds (But if `buildPlatform` is cached, we have a better situation)

**Comment by sky1e** [comment]
https://github.com/NixOS/nixpkgs/blob/1207d1b4ac411eda37b97d91b2d535e20a6cd6fd/pkgs/build-support/setup-hooks/install-shell-files.sh#L92-L95
The `installShellCompletion` command documentation should be updated to not recommend this.

**Comment by Princemachiavelli** [comment]
I think we could also run the targetPlatform binary in a user-mode qemu wrapper? It's very slow but probably enough for just generating the shell auto-completion or man pages.

**Comment by SFrijters** [comment]
Just found this issue while doing some random package fixups after https://github.com/NixOS/nixpkgs/pull/289517 : failing to generate non-empty completions will cause a build failure, so adding a guard for cross-compilation becomes more necessary.

Some extra discussion about how to possibly handle this happened in https://github.com/NixOS/nixpkgs/pull/316749 .

**Comment by jcaesar** [comment]
Noticed this while working on #323161.

Looking around a bit,
```console
$ rg -lU 'installShellCompletion(?s).*<\(\$out/bin' | wc -l
329
```
is definitely more than the handful @SuperSandro2000 predicted [here](https://github.com/NixOS/nixpkgs/pull/289517#pullrequestreview-2092500881). (Looking at a small random sample, it seems most packages don't guard against this at all.)

One more "uuuh" I noticed is that so far, without the check, builds like `github:NixOS/nixpkgs/nixpkgs-24.05#pkgsCross.aarch64-multiplatform.qrtool` would succeed on both hosts with and without the qemu-binfmt setup, but produce different results. Awkward. So disabling the check for cross compilation also isn't a good idea.

**Comment by Aleksanaa** [comment]
> One more "uuuh" I noticed is that so far, without the check, builds like `github:NixOS/nixpkgs/nixpkgs-24.05#pkgsCross.aarch64-multiplatform.qrtool` would succeed both on hosts with and without the [qemu-binfmt](https://search.nixos.org/options?channel=24.05&show=boot.binfmt.emulatedSystems&from=0&size=50&sort=relevance&type=packages&query=boot.binfmt.emulatedSystems) setup, but produce different results.

What do you think of https://github.com/NixOS/nixpkgs/pull/308055#discussion_r1586598623? It's more complicated but won't have this problem

**Comment by SFrijters** [comment]
For what it's worth, I like that `buildPackages` approach, although that would be a lot of manual work to fix up and I'm not sure if I'm missing any footguns/side effects.

**Comment by jcaesar** [comment]
@Aleksanaa I don't think my thoughts on this matter. But for what it's worth: I don't think shell completions or man pages matter enough when cross compiling to warrant this complexity. [This](https://github.com/NixOS/nixpkgs/pull/316749/files) kind of change is the largest amount of complexity I'd be willing to invest. In fact, I'm preparing a PR that does this to a larger batch of packages (around 70).  
If anything, I'd like to see a generic solution that prevents any kind of accidental execution of target binaries in a host build, to avoid reproducibility problems from binfmt_misc. But I can only think of ptrace hacks (or patching bash, if we only want to catch this being done from nix build scripts. Or VMs.), not ideal.

**Comment by SFrijters** [comment]
> If anything, I'd like to see a generic solution that prevents any kind of accidental execution of target binaries in a cross build, to avoid reproducibility problems from binfmt_misc. But I can only think of ptrace hacks (or patching bash, if we only want to catch this being done from nix build scripts. Or VMs.), not ideal.

Yes that would be nice to have. I was very surprised that some cross builds were working while I expected them to fail, only to remember that I have `binfmt.emulatedSystems = [ "aarch64-linux" ];` set on my build machine. I don't know about checking it on the shell script side, but I find it curious that this kind of system config leaks into the build at all.

**Comment by Aleksanaa** [comment]
> But I can only think of ptrace hacks (or patching bash, if we only want to catch this being done from nix build scripts. Or VMs.), not ideal.

Someone once mentioned that additional elf flag bits can be used to prevent the execution of packages prepared for `hostPlatform`, even in native compilation. Maybe you could search for the discussions about `strictDeps`.

**Comment by jcaesar** [comment]
#326243 fixes a chunk of the problems. @SFrijters Any thoughts on just going by foot all the way like that?

**Comment by SFrijters** [comment]
Even though it's a lot of work (I started fixing and verifiying some packages myself, but no bulk effort), I think grinding through may be the best way to just get it done. I do wonder if it may be worthwhile to make a function/alias for this specific conditional in this specific case, to make it easier to grep/change in case it is decided that we do want a different check/implementation for shell completions.

**Comment by Princemachiavelli** [comment]
I think a usermode qemu wrapper would fix more cases and is an established build practice; cmake provides the `CMAKE_CROSSCOMPILING_EMULATOR` option specifically so cmake builds can execute target machine binaries.

A minimal POC:
```nix
cross = legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform;
qemu = cross.buildPackages.qemu.override { nixosTestRunner = true; } # to ignore gui libs, could be even smaller.
qemuRun = "${qemu}/bin/qemu-${t.hostPlatform.qemuArch}";
qemuRunIf = lib.optionalString (stdenv.hostPlatform != stdenv.buildPlatform) qemuRun;
...
installShellCompletion --cmd gum --bash <(${qemuRunIf} $out/bin/gum completion bash)
```

**Comment by alyssais** [comment]
Copying the documentation from buildPackages is definitely the better way to go here.  Emulation is a hack that brings in a large closure and is only supported between some combinations of platforms.  It should be avoided unless absolutely necessary (an example where there's no alternative is gobject-introspection), and this isn't one of those situations.
